### PR TITLE
Add table of branch comparisons

### PR DESCRIPTION
**Pull request by chapulina**
I've been using a [simple gist](https://gist.github.com/chapulina/b93db71fb2c5d97282af2e02bbf5fb81) to keep track of what branches need to be merged forward. Since this can be handy to all maintainers, I'm moving it to a table on this repository.

Also fixed some other things I noticed:

* Remove `markdown-header-`, because all links work with or without it
* Change `master` to `main` where needed
* Tweaked some things under the "Process" section to reflect what we've actually been doing

### CODE REVIEW

File: contributing.md
Comment (adlarkin) [inline contributing.md:389]:
```suggestion
* Create a "merge commit” when porting changes forward, and "Rebase and merge" when porting backwards.
```

File: contributing.md
Comment (adlarkin) [inline contributing.md:400]:
```suggestion
* See [this list](./tools/branch_comparisons.md) to check if a branch needs porting.
```

Personally, I prefer relative paths instead of GitHub links in case things like the branch name ever changes (for example, this is targeting `master` right now, but I am assuming that this will be switched to `main` later).

File: contributing.md
Comment (adlarkin) [inline contributing.md:427]: With the comma between squash and rebase, I found this sentence to be a little confusing - does it mean "do not squash, rebase instead" or does it mean "do not squash or rebase"? Perhaps we can modify this to make it clearer.

File: contributing.md
Comment (adlarkin) [inline contributing.md:416]:
```suggestion
* In the rare event that a pull request needs to be backported (i.e. from a
  higher version to a lower version), use `git cherry-pick` instead of `git merge`, for example:
```

I'm not sure if this is correct or not, but this is how I interpret it. If this suggestion is correct, I think it would be worth applying because the original backport instructions left me confused as to whether I should run `git merge` at all or not.

File: tools/branch_comparisons.md
Comment (adlarkin) [inline tools/branch_comparisons.md:17]:
```suggestion
Library | B ➡️  C | C ➡️  D | D ➡️  main
```
When I looked at this table originally, it wasn't super clear to me what version of each repository was being compared to `main` until I clicked on the `➡️  main` link. It looks like the `➡️  main` column is comparing the most recent ignition release (in this case, `Dome`) to `main`, which makes sense - perhaps we should explicitly state this in the table so it's clear to all readers?

File: tools/branch_comparisons.md
Comment (adlarkin) [inline tools/branch_comparisons.md:6]:
```suggestion
If a cell in the table is `-`, this means that the two Ignition releases use the same version of this Ignition library.
So, for example, `ign-cmake` has a `-` for B ➡️  C because Blueprint and Citadel use `ign-cmake2`.

When it says "There isn’t anything to compare," this means no pull requests
need to be forward-ported.
```

File: tools/branch_comparisons.md
Comment (adlarkin) [inline tools/branch_comparisons.md:33]: Do we want to make any mention of `sdformat`, or does that not apply here since that's not a part of the Ignition org?

**Review by adlarkin — COMMENTED** [review]
This looks pretty good to me. I have a few small comments/questions.

**Comment by adlarkin** [comment]
A few other thoughts:
1. I noticed that this PR is targeting `master` - are we planning to change the default branch of this repository to `main` at some point?
2. Does the contributing guide here somehow sync up with the [contributing guide on the ignitionrobotics website](https://ignitionrobotics.org/docs/all/contributing)? It would be nice to keep the two synced somehow.
3. Some of the Ignition GitHub repositories link to the contributing guide on [ignitionrobotics.org](https://ignitionrobotics.org/home) instead of linking to the contributing guide in this repo (for example, [ign-gazebo](https://github.com/ignitionrobotics/ign-gazebo/blob/main/CONTRIBUTING.md)). Should we make all Ignition repositories point to the contributing guide here on the docs repo instead of the ignitionrobotics website?

File: contributing.md
Comment (chapulina) [reply contributing.md:389]: The button says "Create a merge commit" though :wink: 

image

File: contributing.md
Comment (chapulina) [reply contributing.md:400]: So, the problem here is that when this is hosted on https://ignitionrobotics.org/docs/all/contributing, the comparison document is not served by that server. This is convenient because we can update the table on GitHub and don't need to redeploy the web server.

File: contributing.md
Comment (chapulina) [reply contributing.md:427]: I actually meant "do not squash, rebase instead", I'll update

File: contributing.md
Comment (adlarkin) [reply contributing.md:400]: I didn't realize that this was used to host the contributing guide on the ignitionrobotics website - using the full link makes sense then. I'll mark this as resolved.

File: tools/branch_comparisons.md
Comment (chapulina) [reply tools/branch_comparisons.md:37]: The reason I left SDFormat out is because that repository hasn't been using the same merge strategy. Changes are being cherry-picked then rebase-merged on all directions (forwards and backwards). That makes the GitHub comparison a bit crazy. For example see the diff from `sdf8` to `sdf9`:

https://github.com/osrf/sdformat/compare/sdf9...sdf8

That shows for example as if `Material::FilePath` wasn't on `sdf9` yet. Which is not true, because it's there :woman_shrugging: :

https://github.com/osrf/sdformat/blob/sdf9/src/Material.cc#L355-L359

@scpeters , you've been cherry-picking those changes. How do you keep track of what needs porting? It would be good to document that here too so all maintainers can help out. Ideally, we'd use the same strategy for all libraries, but I understand SDFormat has been following a slightly different process.

File: contributing.md
Comment (chapulina) [reply contributing.md]: 824176b170a8353a14f69bde100af7982fd8aa53

File: contributing.md
Comment (chapulina) [reply contributing.md]: 824176b170a8353a14f69bde100af7982fd8aa53

File: tools/branch_comparisons.md
Comment (chapulina) [reply tools/branch_comparisons.md]: Good point, I don't remember exactly what my reasoning was to left that out but I think that makes sense. Changed in 824176b170a8353a14f69bde100af7982fd8aa53

File: tools/branch_comparisons.md
Comment (chapulina) [reply tools/branch_comparisons.md:10]: 824176b170a8353a14f69bde100af7982fd8aa53

**Comment by chapulina** [comment]
> I noticed that this PR is targeting master - are we planning to change the default branch of this repository to main at some point?

Yeah this repository is in the batch that will be migrated on January

> Does the contributing guide here somehow sync up with the contributing guide on the ignitionrobotics website?

Yup, that guide is pulled directly from here whenever we deploy the web server (which is a manual step for now).

> Should we make all Ignition repositories point to the contributing guide here on the docs repo instead of the ignitionrobotics website?

The link on `ignitionrobotics.org` is the canonical location. This repository should be treated as the source code for that site and never be referenced directly. It would be good to make this clearer to contributors. Feel free to increment the README as you see fit :wink:

File: tools/branch_comparisons.md
Comment (scpeters) [reply tools/branch_comparisons.md:37]: Yes, I've been cherry-picking instead of merging. I use the following branches to keep track of outdated changes:

* https://github.com/osrf/sdformat/tree/merge_8_9_reference
* https://github.com/osrf/sdformat/tree/merge_9_10_reference

I regularly merge `sdf{X}` and `sdf{X+1}` into these branches and resolve conflicts. To see what needs to be merged forward from `sdf9` to `sdf10` for example:

1. Make sure that you've merged forward from `sdf9` to `merge_9_10_reference` by ensuring that [compare/merge_9_10_reference...sdf9](https://github.com/osrf/sdformat/compare/merge_9_10_reference...sdf9) is empty; if not, then merge `sdf9` into that branch.

2. Check to see what is in `merge_9_10_reference ` that hasn't yet been merged forward to `sdf10`, by looking at the diff of [compare/sdf10...merge_9_10_reference](https://github.com/osrf/sdformat/compare/sdf10...merge_9_10_reference#files_bucket) (the list of commits isn't helpful).

At the moment that I type this, both of those comparisons are not empty, meaning that there are changes from `sdf9` that need to be merged forward to `merge_9_10_reference`, and changes in `merge_9_10_reference` that aren't yet in `sdf10`.

I'm not sure if other people like this workflow or not, but it's kept the sdformat release branches pretty clean so far.

Comment (chapulina) [reply tools/branch_comparisons.md:42]: I took a stab at documenting SDF's case in 2c1b8e88697fa8376fcafeb1accbe854f6e8b7a0. Feel free to make edits directly to this branch to improve it.

> I'm not sure if other people like this workflow or not, but it's kept the sdformat release branches pretty clean so far.

I must say I'm not convinced that this workflow is scalable and maintainable in the long run. There are lots of moving pieces to be kept in sync, and as you mentioned, the diff commit list isn't helpful. Sure, it keeps a linear history, but at a high maintenance cost. Is that worth it? It may be suitable for a low-traffic repository that has someone keeping a close eye at every change, but I think it will become more difficult as hundreds of changes get intertwined across different branches. I think it also makes it more difficult to share the responsibility across more people in the team.

That said, I'm not opposed to keeping the status quo a bit longer to see how it goes. I think it's good to do some experimentation. The process we're doing for the Ignition libraries is far from perfect. There are extra merge commits sprinkled all over the history, and it's harder to track what was merged when. I'm all ears for improvements. But in my opinion, so far, that's the option that scales better across several libraries with several versions and several maintainers.

File: tools/branch_comparisons.md
Comment (adlarkin) [inline tools/branch_comparisons.md:23]:
```suggestion
> SDFormat is a special case that doesn't follow the forward-port process listed in the contributing guide.
  For SDFormat, the release branches should be compared against `merge_N_M_reference` branches (where `M = N + 1`) instead of other release branches.
  Be sure to update the `merge_N_M_reference` branches before using them for comparison.
  The `merge_N_M_reference` branches need to be updated if `compare/merge_N_M_reference...sdfN` isn't empty.
  If this comparison isn't empty, update `merge_N_M_reference` by merging `sdfN` into `merge_N_M_reference`.
  After updating `merge_N_M_reference`, see what's in `merge_N_M_reference` that hasn't been merged forward to `sdfM` yet by looking at the **diff** of `compare/sdfM...merge_N_M_reference` (the list of commits in this comparison isn't helpful), and cherry-pick accordingly.
```

I tried to clarify this a little more based on this [comment](https://github.com/ignitionrobotics/docs/pull/108#discussion_r509484526), but I'm not 100% sure I understand how the cherry-picking part works - do we cherry-pick commits into a separate branch that is then used to create a forward-port PR?

File: tools/branch_comparisons.md
Comment (adlarkin) [inline tools/branch_comparisons.md:13]:
```suggestion
Otherwise, those branches could use a forward-port pull request.
See the instructions on the
```

**Review by adlarkin — COMMENTED** [review]
Most of the Ignition stuff here looks good, but I had some comments about SDFormat.

Also, I noticed that we misspelled `under` in point number 4 of the [writing tests section](https://github.com/ignitionrobotics/docs/blob/59055ade9b5edf65c0ebb2c74a93613e976452c9/contributing.md#writing-tests) (performance tests) in `contributing.md` - the last part of this sentence says, `go unde tests/performance`, when it should really say `go under tests/performance`. We should update that typo as well (unfortunately, I wasn't able to leave a comment on that line of the changed files when looking at the diff).

File: tools/branch_comparisons.md
Comment (scpeters) [reply tools/branch_comparisons.md:42]: after talking about it with @azeey, I'm giving up the cherry-pick pattern with sdformat and have opened a PR to start merging forward from sdf8 -> sdf9 and will continue with the other branches: https://github.com/osrf/sdformat/pull/445

so sdformat can be treated like the others with consistency

Comment (chapulina) [reply tools/branch_comparisons.md:42]: Thanks for iterating, @scpeters . Updated in 81dd4f674d0fc7f9a78667760e3becf281a45ce4.

File: tools/branch_comparisons.md
Comment (chapulina) [reply tools/branch_comparisons.md:23]: SDFormat is using the same pattern as the rest now, so this isn't needed.

**Comment by chapulina** [comment]
I think all comments have been addressed. Moving this back to in review.

File: tools/branch_comparisons.md
Comment (adlarkin) [inline tools/branch_comparisons.md:9]:
```suggestion
When the comparison link says "There isn’t anything to compare," this means no pull requests
```

"it" is a little vague here, so I think it would be good to clarify that we are referring to the branch comparison, and not the table itself.

File: tools/branch_comparisons.md
Comment (adlarkin) [inline tools/branch_comparisons.md:23]: Do we still need this if SDFormat is following the Ignition porting strategy now?

**Review by adlarkin — APPROVED** [review]
Just a few comments before merging. Also, I fixed a typo in a different part of the contributing guide in https://github.com/ignitionrobotics/docs/pull/108/commits/59beda63e864779374f08fab94b8a61a0a6ea37e.

File: tools/branch_comparisons.md
Comment (chapulina) [reply tools/branch_comparisons.md]: a667d98

File: tools/branch_comparisons.md
Comment (chapulina) [reply tools/branch_comparisons.md]: a667d98

**Pull request merged**
